Split ES controller

caseydavenport opened the pull request:
## Description

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

The ES controller had gotten big and complicated. This PR splits it into several sub controllers which are smaller and
more focused in scope.

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

caseydavenport committed:
Prototype ability for operator to install multiple copies of
tigera-manager

caseydavenport committed:
Prototype ability for operator to install multiple copies of
tigera-manager

caseydavenport committed:
Multi-tenant support

Josh-L committed:
-Add basic multi-tenancy Manager UTs
-Adding multi-tenant reconciliation test to manager_controller
as well as a few other minor Manager fixups
-Multi-tenantize PolicyRecommendation
-Add new tenancy.go in new "tenancy" package intended for multi-tenant
utility functions. Replace Manager and PolicyRecommendation controller
blocks with call to new utility function

dimitri-nicolo committed:
[EV-3840] Add networkpolicies to policy rec clusterrole

caseydavenport committed:
Merge pull request #2715 from caseydavenport/casey-mt-prototype

Prototype ability for operator to install multiple copies of tigera-manager

rene-dekker committed:
Postpone policy recommendation scope watches until apiserver is up an… (#2713)

* Postpone policy recommendation scope watches until apiserver is up and running

* Block the policy recommendation controller if the apiserver watch is not established yet

* Mark readyFlag in unit tests.

tmjd committed:
Compliance: Add watch for ES PubliceCertSecret

tmjd committed:
Merge pull request #2718 from tmjd/compliance-add-es-pub-cert-secret-watch

Compliance: Add watch for ES PublicCertSecret

electricjesus committed:
fix: WAF rules should be intialized as DetectionOnly post-install (#2716)

caseydavenport committed:
Merge pull request #2721 from caseydavenport/casey-rebase

Rebase on latest master

marvin-tigera committed:
Merge pull request #2714 from dimitri-nicolo/dimitri-EV-3840

[EV-3840] Add networkpolicies to policy rec clusterrole

anthonytwh committed:
fix typo in filename to pod_security_policy.go

Josh-L committed:
Fix "make fv" target from incorrectly running UTs. Also remove
non-existent "vendor" directory skipPackage directive.

marvin-tigera committed:
Merge pull request #2725 from anthonytwh/psp-filename

fix typo in filename to pod_security_policy.go

tmjd committed:
Revert "Update k8s dependencies to 1.26 release train"

This reverts commit a3b32bc63a48057586f6aab5e793f7421e681419.

tmjd committed:
Add back Makefile change from reverted commit

From the original PR
> add the -s flag to reduce binary size / strip symbols

* Also fix Copyright in file changed in revert
* Update CRDs since k8s version changed

tmjd committed:
Merge pull request #2727 from tmjd/revert-k8s126-update

Revert k8s126 update

rene-dekker committed:
Update ES&KB to v7.17.11. (#2730)

Josh-L committed:
Point UT and FV test invocations at their respective start directories instead of filtering out unwanted directories

caseydavenport committed:
Merge pull request #2734 from caseydavenport/casey-rebase

Rebase on latest master

Behnam-Shobiri committed:
Update golang to 1.20.6

rene-dekker committed:
make gen-files (#2737)

Behnam-Shobiri committed:
Merge branch 'tigera:master' into ben-update-go-ma

marvin-tigera committed:
Merge pull request #2735 from Behnam-Shobiri/ben-update-go-ma

Update golang to 1.20.6

Brian-McM committed:
Merge pull request #2726 from Josh-Tigera/josh.fvs

Fix "make fv" target from incorrectly running UTs. Also remove

asincu committed:
Use internal manager tls certificate for es-proxy and linseed (#2739)

communication

rene-dekker committed:
Don't use subPath on windows: https://github.com/moby/moby/issues/30555 (#2740)

Josh-L committed:
CI reporting improvements (#2752)

* CI reporting improvements including:
-Silencing Ginkgo deprecation/new version noise
-Adding stacktrace reporting for all failures
-Restructure reports directory into ut/fv and publish each accordingly

* Add pipeline-level results report

tmjd committed:
Add watch for managed cluster linseed cert in compliance

tmjd committed:
Merge pull request #2753 from tmjd/compliance-watch-linseed

Add watch for managed cluster linseed cert in compliance

Juneezee committed:
Replace `github.com/ghodss/yaml` with `sigs.k8s.io/yaml`

At the time of making this commit, the package `github.com/ghodss/yaml`
is no longer actively maintained.

`sigs.k8s.io/yaml` is a permanent fork of `ghodss/yaml` and is actively
maintained by Kubernetes SIG.

Signed-off-by: Eng Zer Jun <engzerjun@gmail.com>

hjiawei committed:
Add PSPs for dex and policy recommendation

PSPs are needed in a k8s <= v1.24 cluster when pod admission controller is
enabled. This change adds PSPs for Tigera dex and policy recommendation
components.

hjiawei committed:
Add .PHONY for ut target and ignore ut folder

tmjd committed:
Merge pull request #2758 from hjiawei/make-ut-run-again

Add .PHONY for ut target and ignore ut folder

Juneezee committed:
Update copyright

Signed-off-by: Eng Zer Jun <engzerjun@gmail.com>

marvin-tigera committed:
Merge pull request #2755 from Juneezee/chore/yaml

Replace `github.com/ghodss/yaml` with `sigs.k8s.io/yaml`

MichalFupso committed:
Openshift PodSecurity (#2724)

* openshift security context

* openshift security context

* node security context

* egw podsecurity

* render tests

ti-afra committed:
update CRD (felixConfiguration)

ti-afra committed:
update CRD (felixConfiguration) (#2759)

marvin-tigera committed:
Merge pull request #2757 from hjiawei/psp-dex-polrec

Add PSPs for dex and policy recommendation

ti-afra committed:
update CRD (OSS felixConfiguration)

Brian-McM committed:
Merge pull request #2765 from ti-afra/fix-tag

update CRD (OSS felixConfiguration)

caseydavenport committed:
Recreate certificates that are only specified to be used as server certs

- Only require key usage for certain secret certs
- Compliance: order dependencies so certificates are updated before pods
  are updated.

Josh-L committed:
FV test improvements including:
-Add localized caching of OSS Calico images to speed up performance and
improve test reliability by removing variability of image pull timing
per test
-Fix race condition in CRD management tests
-Fix race condition in mainline tests
-Use non-caching k8s client and remove use of unstructured types

Josh-L committed:
FV test improvements including:
-Add localized caching of OSS Calico images to speed up performance and
improve test reliability by removing variability of image pull timing
per test
-Fix race condition in CRD management tests
-Fix race condition in mainline tests
-Use non-caching k8s client and remove use of unstructured types

caseydavenport committed:
Merge remote-tracking branch 'origin/master' into feature-mt

caseydavenport committed:
Merge pull request #2771 from caseydavenport/casey-merge-master

Merge in latest master

Josh-L committed:
PR feedback changes

Josh-L committed:
PR feedback changes

Josh-L committed:
Upload test results even for failures

Josh-L committed:
Upload test results even for failures

Josh-L committed:
More PR feedback changes

Josh-L committed:
More PR feedback changes

Josh-L committed:
Fix go vet

Josh-L committed:
Fix go vet

Josh-L committed:
Formatting fixes

Josh-L committed:
Formatting fixes

Josh-L committed:
Merge pull request #2772 from Josh-Tigera/josh.flake-fixes

-Add localized caching of OSS Calico images to speed up performance and improve test reliability by removing variability of image pull timing per test. Also has the added benefit of not ripping through docker pull request limits
-Fix race condition in CRD management tests
-Fix race condition in mainline tests
-Use non-caching k8s client and remove use of unstructured types

Josh-L committed:
Merge remote-tracking branch 'origin/master' into josh.master-merge

Josh-L committed:
Fix typo

Josh-L committed:
Merge branch 'josh.master-merge' of github.com:Josh-Tigera/operator into josh.master-merge

Josh-L committed:
Fixup tests to incorporate multi-tenant changes

Josh-L committed:
Fixup UT and FVs to get them passing

Josh-L committed:
Get around ineffeassign linter bug

tmjd committed:
Merge pull request #2775 from tmjd/cert-fix

Cert fix

Josh-L committed:
Merge pull request #2774 from Josh-Tigera/josh.master-merge

Merge upstream master branch

sridhartigera committed:
Fix for EGW deployment in openshift 4.13

mgleung committed:
Merge pull request #2779 from sridhartigera/openshift-egw

Fix for EGW deployment in openshift 4.13

Josh-L committed:
Fix UTs

tmjd committed:
Certs: Returning specific error type when key usage is wrong

tmjd committed:
Logstorage: do not let unupdated key usage block progress

hjiawei committed:
Add SecurityContext for Prometheus service container (#2784)

caseydavenport committed:
Merge pull request #2723 from Josh-Tigera/josh-mt-prototype

Multi-tenant changes for PolicyRecommendation

tmjd committed:
Merge pull request #2790 from tmjd/refix-certs

Ensure logstorage controller is not blocked on certs that need to be updated by other controllers

caseydavenport committed:
Merge remote-tracking branch 'origin/master' into casey-merge-master

caseydavenport committed:
Merge pull request #2792 from caseydavenport/casey-merge-master

Merge master

rene-dekker committed:
Remove duplicate Felix Health Port for Openshift

caseydavenport committed:
UT fixes for multi-tenant refactor

marvin-tigera committed:
Merge pull request #2795 from tigera/ev-4017

Remove duplicate Felix Health Port for Openshift

rene-dekker commented on pkg/controller/applicationlayer/applicationlayer_suit_test.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+// Copyright (c) 2021, 2023 Tigera, Inc. All rights reserved.
```
Just noticed that the File name has a typo

rene-dekker commented on pkg/controller/certificatemanager/certificatemanager.go:
```diff
@@ -365,3 +435,53 @@ func (cm *certificateManager) CreateTrustedBundle(certificates ...certificateman
 	return certificatemanagement.CreateTrustedBundleWithSystemRootCertificates(append([]certificatemanagement.CertificateInterface{cm.keyPair}, certificates...)...)
 }
+
+func (cm *certificateManager) LoadTrustedBundle(ctx context.Context, client client.Client, ns string) (certificatemanagement.TrustedBundleRO, error) {
```
In need of a comment and unit test.

rene-dekker commented on pkg/controller/logstorage/elastic/elastic_controller.go:
```diff
@@ -0,0 +1,710 @@
+
+	// Watch all the secrets created by this controller so we can regenerate any that are deleted
+	for _, secretName := range []string{
+		render.TigeraElasticsearchGatewaySecret, render.TigeraKibanaCertSecret,
+		render.OIDCSecretName, render.DexObjectName, esmetrics.ElasticsearchMetricsServerTLSSecret,
+		render.TigeraLinseedSecret,
```
There are secretwatches that can be added to this loop.

CASecretName is missing

rene-dekker commented on pkg/controller/logstorage/elastic/elastic_controller.go:
```diff
@@ -0,0 +1,710 @@
+
+	// Wait for the initializing controller to indicate that the LogStorage object is actionable.
+	if ls == nil || ls.Status.State != operatorv1.TigeraStatusReady {
+		r.status.SetDegraded(operatorv1.ResourceNotReady, "Waiting for LogStorage defaulting to occur", nil, reqLogger)
```
For managed clusters (that have no logstorage), would this message be misleading? And probably we would not want to requeue

rene-dekker commented on pkg/controller/logstorage/elastic/elastic_controller.go:
```diff
@@ -0,0 +1,710 @@
+		return reconcile.Result{}, err
+	}
+
+	// If this is a Managed cluster ls must be nil to get to this point (unless the DeletionTimestamp is set) so we must
```
We won't get to this point . See comment ^

rene-dekker commented on pkg/controller/logstorage/initializer/initializing_controller.go:
```diff
@@ -0,0 +1,304 @@
+	}
+
+	// Configure watches for operator.tigera.io APIs this controller cares about.
+	if err = c.Watch(&source.Kind{Type: &operatorv1.LogStorage{}}, &handler.EnqueueRequestForObject{}); err != nil {
```
I think this controller is also dependent on Installation{} and should watch it.

hjiawei committed:
Increase probes timeout for calico components (#2746)

* Increase probes timeout for calico components

This changeset increases k8s liveness and readiness probe period and
timeout for various Calico components. Before this change, most of
the components are using k8s default `TimeoutSeconds` (1s) and
`PeriodSeconds` (10s). These defaults cause some components to become
unhealthy quickly and got restarted by k8s engine, especially under
a busy cluster or with long running tasks.

We have tuned period and timeout seconds as follows:
* liveness probe: bump timeout to 5s and period to 60s. One component
  will be restarted after it fails to response for 3 minutes (with
  default `FailureThreshold` 3).
* readiness probe: bump timeout to 5s and period to 30s. One component
  will be removed from service after 1.5 minute (with default
  `FailureThreshold` 3).
* certain components (like fluentd or elasticsearch) will have a longer
  timeout, period, and/or failure threshold.

* Move probe default settings in component machinery

* Add UTs for probe default values

* Rename function and fix a few code formatting

* Use default timeout for ids controller probe

* Add a comment for egw initial delay increase

rene-dekker committed:
Create internal manager tls inside of manager_controller and fix certificate management.

caseydavenport committed:
Merge remote-tracking branch 'origin/master' into casey-merge-master

caseydavenport committed:
Merge pull request #2803 from caseydavenport/casey-merge-master

Merge master

rene-dekker committed:
Allow certificate management w/ management clusters

marvin-tigera committed:
Merge pull request #2806 from rene-dekker/automated-cherry-pick-of-#2802-origin-master

Automated cherry pick of #2802: Create internal manager tls inside of manager_controller

asincu committed:
Internal-manager-tls should contain only K8S services (#2810)

In order for Voltron (or the reverse proxy inside Voltron) to perform SNI, we must provide two sets of certificates that do not have the same SANs domains. Manager-tls will be used for external communication (requests originating from the browser) and internal-manager-tls will be used for inter-cluster communications.

hjiawei committed:
Use default non-root SecurityContext for manager (#2809)

rene-dekker committed:
Fix watch that is monitoring the wrong namespace (#2814)

electricjesus committed:
fix policyrecommendationscopes failing dirty-check (#2818)

electricjesus committed:
go generate pkg/render/applicationlayer (#2816)

* go generate pkg/render/applicationlayer

* make fix

caseydavenport committed:
Merge remote-tracking branch 'origin/master' into feature-mt

caseydavenport committed:
Merge pull request #2820 from caseydavenport/casey-merge-master

Merge master

caseydavenport committed:
Fix managed clusters - split out to new controller for logstorage resources

caseydavenport committed:
Fix tests after log storage managed cluster change

caseydavenport committed:
Run make fix
